src/Catalyst.KBucket/KBucket.cs

```diff
@@ -0,0 +1,528 @@
+        /// <param name="candidate">
+        ///   The new contact.
+        /// </param>
+        /// <returns>
+        ///   Always returns the <paramref name="incumbent"/>.
+        /// </returns>
+        public static T DefaultAbiter(T incumbent, T candidate) { return incumbent; }
+
+        /// <summary>
+        ///   Finds the XOR distance between the two contacts.
+        /// </summary>
+        public UInt256 Distance(T a, T b)
+        {
+            Validate(a);
+            Validate(b);
+            return Distance(a.Id, b.Id);
+        }
+
+        /// <summary>
+        ///   Finds the XOR distance between the two contact IDs.
+        /// </summary>
+        public UInt256 Distance(byte[] a, byte[] b)
+        {
+            UInt256 distance = 0;
+            var i = 0;
+            var min = Math.Min(a.Length, b.Length);
+            var max = Math.Max(a.Length, b.Length);
+            for (; i < min; ++i) distance = distance * 256 + ((ulong) a[i] ^ b[i]);
+
+            for (; i < max; ++i) distance = distance * 256 + 255;
+
+            return distance;
+        }
+
+        /// <summary>
+        ///   Gets the closest contacts to the provided contact.
+        /// </summary>
+        /// <param name="contact"></param>
+        /// <returns>
+        ///   An ordered sequence of contact, sorted by closeness. 
+        /// </returns>
+        /// <remarks>
+        ///   "Closest" is the XOR metric of the contact.
+        /// </remarks>
+        public IEnumerable<T> Closest(T contact)
+        {
+            Validate(contact);
+            return Closest(contact.Id);
+        }
+
+        /// <summary>
+        ///   Gets the closest contacts to the provided contact.
+        /// </summary>
+        /// <param name="id">
+        ///   The unique <see cref="IContact.Id"/> of a contact.
+        /// </param>
+        /// <returns>
+        ///   An ordered sequence of contact, sorted by closeness. 
+        /// </returns>
+        /// <remarks>
+        ///   "Closest" is the XOR metric of the contact.
+        /// </remarks>
+        public IEnumerable<T> Closest(byte[] id)
+        {
+            return this.Select(c => new
+                {
+                    distance = Distance(c.Id, id), contact = c
+                })
+               .OrderBy(a => a.distance).Select(a => a.contact);
+        }
+
+        /// <inheritdoc />
+        public int Count => Root.DeepCount();
+
+        /// <inheritdoc />
+        public bool IsReadOnly => false;
+
+        /// <inheritdoc />
+        public void Add(T item)
+        {
+            Validate(item);
+            bool q;
+            PingEventArgs<T> e;
+            _rwlock.EnterWriteLock();
+            try
+            {
+                q = _Add(item, out e);
+            }
+            finally
+            {
+                _rwlock.ExitWriteLock();
+            }
+
+            // Could not add.  Ping oldest contacts.
+            if (!q) Ping?.Invoke(this, e);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)